clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -217,31 +259,60 @@ class device_vector {
   T *real_begin() { return _storage; }
   const T *real_begin() const { return _storage; }
   void swap(device_vector &v) {
-    ::std::swap(_size, v._size);
-    ::std::swap(_capacity, v._capacity);
-    ::std::swap(_storage, v._storage);
-    ::std::swap(_alloc, v._alloc);
+    if constexpr (::std::allocator_traits<
+                      Allocator>::propagate_on_container_swap::value) {
+      ::std::swap(_alloc, v._alloc);
+      ::std::swap(_size, v._size);
+      ::std::swap(_capacity, v._capacity);
+      ::std::swap(_storage, v._storage);
+    } else {
+      // swap all elements up to the minimum size between the two vectors
+      size_type min_size = ::std::min(size(), v.size());
+      for (size_type i = 0; i < min_size; i++) {
+        ::std::swap(_storage[i], v._storage[i]);
+      }
+      // then copy the elements beyond the end of the smaller list, and resize
+      if (size() > v.size()) {
+        size_type tmp_v_size = v.size();
+        v.resize(size());
+        ::std::copy(
+            oneapi::dpl::execution::make_device_policy(get_default_queue()),
+            begin() + tmp_v_size, end(), v.begin() + tmp_v_size);
+        resize(tmp_v_size);
+      } else {
+        size_type tmp_size = v.size();
+        resize(v.size());
+        ::std::copy(
+            oneapi::dpl::execution::make_device_policy(get_default_queue()),
+            v.begin() + tmp_size, v.end(), begin() + tmp_size);
+        v.resize(tmp_size);
+      }
+    }
   }
   reference operator[](size_type n) { return _storage[n]; }
   const_reference operator[](size_type n) const { return _storage[n]; }
   void reserve(size_type n) {
     if (n > capacity()) {
       // allocate buffer for new size
-      auto tmp = _alloc.allocate(2 * n);
+      auto tmp = alloc_traits::allocate(_alloc, n);
       // copy content (old buffer to new buffer)
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                begin(), end(), tmp);
-      // deallocate old memory
-      _alloc.deallocate(_storage, _capacity);
+      if (capacity() > 0) {
+        ::std::copy(
+            oneapi::dpl::execution::make_device_policy(get_default_queue()),
```

Comment:
I think here we can not use dpl to speed up, as construction is doing here.

Reply:
We are doing a copy to move existing constructed data to a new memory location, not calling the constructor here.  I don't believe any constructor is called here, and thus, using oneDPL to accelerate the copy is appropriate.

Reply:
The new memory is not initialized, for some complex type, it may cause unexpected errors when doing copy assignment.